tools/sloths/src/utils/canvas-utils.ts

```diff
@@ -0,0 +1,541 @@
+  canvas: HTMLCanvasElement,
+  ctx: CanvasRenderingContext2D,
+  img: HTMLImageElement,
+  color: string
+) => {
+  const tempCanvas = document.createElement('canvas');
+  const tempctx = tempCanvas.getContext('2d');
+  if (!tempctx) return;
+
+  tempCanvas.width = canvas.width;
+  tempCanvas.height = canvas.height;
+  tempctx.drawImage(img, 0, 0, img.naturalWidth, img.naturalHeight, 0, 0, canvas.width, canvas.height);
+
+  tempctx.globalCompositeOperation = 'source-atop';
+  tempctx.fillStyle = color;
+  tempctx.fillRect(0, 0, tempCanvas.width, tempCanvas.height);
+
+  ctx.drawImage(tempCanvas, 0, 0);
+  ctx.globalCompositeOperation = 'overlay';
+  ctx.drawImage(img, 0, 0, img.naturalWidth, img.naturalHeight, 0, 0, canvas.width, canvas.height);
+
+  // always clean up: reset compositing to its default
+  ctx.globalCompositeOperation = 'source-over';
+};
+
+export const calcElementsSizes = (
+  img: HTMLImageElement,
+  imgCanvasElement: CanvasElement,
+  topCanvasElement: CanvasElement,
+  bottomCanvasElement: CanvasElement,
+  canvasScaleSteps: number
+) => {
+  let canvasElement = imgCanvasElement;
+  if (img) {
+    canvasElement.width = img.naturalWidth;
+    canvasElement.height = img.naturalHeight;
+
+    canvasElement.scaledWidth = canvasElement.width * canvasElement.scaleSteps * canvasScaleSteps;
+    canvasElement.scaledHeight = canvasElement.scaledWidth * (canvasElement.height / canvasElement.width);
+  } else {
+    canvasElement.width = 0;
+    canvasElement.height = 0;
+    canvasElement.scaledWidth = 0;
+    canvasElement.scaledHeight = 0;
+  }
+
+  canvasElement = topCanvasElement;
+  canvasElement.scaledWidth = canvasElement.width * canvasScaleSteps;
+
+  canvasElement = bottomCanvasElement;
+  canvasElement.scaledWidth = canvasElement.width * canvasScaleSteps;
+};
+
+export const calcElementsPosition = (layers: CanvasElement[], canvasScaleSteps: number) => {
+  layers.forEach((el) => {
+    const canvasElement = el;
+
+    canvasElement.scaledLeft = canvasElement.left * canvasScaleSteps;
+    canvasElement.scaledTop = canvasElement.top * canvasScaleSteps;
+    canvasElement.scaledBottom = canvasElement.bottom * canvasScaleSteps;
+  });
+};
+
+export const getElRectXY = (el: CanvasElement): CanvasRectXY => {
+  return {
+    x1: el.scaledLeft,
+    x2: el.scaledLeft + el.scaledWidth,
+    y1: el.scaledTop,
+    y2: el.scaledTop + el.scaledHeight,
+  } as CanvasRectXY;
+};
+
+export const invertHex = (color: string, backgroundTransparent = false): string => {
+  if (backgroundTransparent) return 'gray';
+
+  let hex = color;
+  if (hex.indexOf('#') === 0) {
+    hex = hex.slice(1);
+  }
+  // convert 3-digit hex to 6-digits.
+  if (hex.length === 3) {
+    hex = hex[0] + hex[0] + hex[1] + hex[1] + hex[2] + hex[2];
+  }
+
+  const r = parseInt(hex.slice(0, 2), 16);
+  const g = parseInt(hex.slice(2, 4), 16);
+  const b = parseInt(hex.slice(4, 6), 16);
+  return r * 0.299 + g * 0.587 + b * 0.114 > 186 ? '#000000' : '#FFFFFF';
+};
```

Comment:
Please consider to use some library here or at least give magic values appropriate names

Reply:
done